internal/sdkio/ringbuffer.go

```diff
@@ -0,0 +1,100 @@
+package sdkio
+
+import (
+	"io"
+)
+
+// RingBuffer struct satisfies io.ReadWrite
+// interface.
+//
+// ReadBuffer is a revolving buffer data structure
+// which can be used to store snapshots of data in a
+// revolving window.
+type RingBuffer struct {
+	slice []byte
+	start int
+	end   int
+	size  int
+}
+
+// New method takes in the capacity as an int
+// and returns a RingBuffer.
+func New(capacity int) *RingBuffer {
+	slice := make([]byte, capacity, capacity)
+	ringBuf := RingBuffer{
+		slice: slice,
+		start: 0,
+		end:   0,
+		size:  0,
+	}
+	return &ringBuf
+}
```

Comment:
Instead of using of `New` creating the byte slice, if the byte slice is passed in it allows the usage of RingBuffer that doesn't allocate the backing array onto the heap.

```go
var buf [1024]byte
ringBuf := sdkio.NewRingBuffer(buf[:])

reader := io.TeeReader(body, ringBuf)

decoder := xml.Decoder(reader)
```

Reply:
[TeeReader](https://golang.org/pkg/io/#TeeReader) takes a reader as first param, and writer as second. So here we can use io.TeeReader(ringBuf, body) as ringBuf satisfies io.ReadWriter; But we can also use a MultiReader in this case. Is that correct?